src/caffeine/program_termination_s.f90

```diff
@@ -62,61 +62,25 @@ end subroutine prif_stop_character
     end if
   end procedure prif_error_stop
 
-  pure subroutine prif_error_stop_character(stop_code)
+  subroutine prif_error_stop_character(stop_code)
     !! stop all images and provide the stop_code as the process exit status
     character(len=*), intent(in) :: stop_code
-
-    interface
-      pure subroutine caf_error_stop_character_c(stop_code, length) bind(C, name = "caf_error_stop_character_c")
-        use, intrinsic :: iso_c_binding, only: c_char, c_int
-        implicit none
-        integer(c_int), intent(in), value :: length
-        character(len=1,kind=c_char), intent(in) :: stop_code(length)
-      end subroutine
-    end interface
-
-    call caf_error_stop_character_c(f_c_string(stop_code), len(stop_code))
-  end subroutine
-
-  subroutine inner_caf_error_stop_character(stop_code, length) bind(C, name = "inner_caf_error_stop_character")
-    integer(c_int), intent(in) :: length
-    character(len=1,kind=c_char), intent(in) :: stop_code(length)
-
     integer(c_int), parameter :: error_occured = 1
 
-    write(error_unit, *) c_f_string(stop_code, length)
+    write(error_unit, *) stop_code
     flush error_unit
 
     call prif_error_stop_integer(error_occured)
-
   end subroutine
 
-  pure subroutine prif_error_stop_integer(stop_code)
+  subroutine prif_error_stop_integer(stop_code)
     !! stop all images and provide the stop_code, or 0 if not present, as the process exit status
     integer, intent(in), optional :: stop_code
-
-    interface
-      pure subroutine caf_error_stop_integer_c(stop_code) bind(C, name = "caf_error_stop_integer_c")
-        use, intrinsic :: iso_c_binding, only: c_int
-        implicit none
-        integer(c_int), intent(in) :: stop_code
-      end subroutine
-    end interface
-
-    call caf_error_stop_integer_c(stop_code)
-  end subroutine
-
-  subroutine inner_caf_error_stop_integer(stop_code) bind(C, name = "inner_caf_error_stop_integer")
-    integer, intent(in), optional :: stop_code
-
     integer exit_code
 
     if (.not. present(stop_code)) then
-
       call caf_decaffeinate(exit_code=1)
-
     else if (stop_code==0) then
-
       write(error_unit) stop_code
```

Comment:
IIUC there's an inconsistency here: the integer error stop code is being written to error_unit only if and only if it's zero. This not part of the spec, and I suspect it arose from the current implementation approach where `prif_error_stop_character` calls `prif_error_stop_integer`.

Of course the actual spec says the `quiet` argument (which is currently ignored by the implementation) controls output behavior, so in that sense both current behaviors are "wrong".

We could consider just fixing #78 along with this PR since it should only require a few extra lines of code in these same subroutines, but I'm also fine with deferring that.

Reply:
Addressed in this PR. Please check the implementation when you are able. I believe this PR should now match the PRIF semantics, including the semantics of the `quiet` arg.